tensorflow_probability/python/experimental/stats/sample_stats.py

```diff
@@ -389,3 +393,185 @@ def finalize(self, state):
+  information, which includes intermediate calculations, are held in a
+  `RunningPotentialScaleReductionState` as returned via `initialize` and
+  `update` method calls.
+
+  `RunningPotentialScaleReduction` is meant to serve general streaming rhat.
+  For a specialized version that fits streaming over MCMC samples, see
+  `RhatReducer` in `tfp.experimental.mcmc`.
+
+  #### References
+
+  [1]: Andrew Gelman and Donald B. Rubin. Inference from Iterative Simulation
+       Using Multiple Sequences. _Statistical Science_, 7(4):457-472, 1992.
+  """
+
+  def __init__(self, shape, num_chains, dtype=tf.float32):
+    """Instantiates this object.
+
+    Args:
+      shape: Python `Tuple` or `TensorShape` representing the shape of
+        incoming samples. Using a collection enables batching and implies
+        that future samples will mimic that exact structure.
+      num_chains: A (possibly nested) collection of integers representing
+        the number of independent chains ran for each sample. Using a
+        collection enables batching and implies that future samples will
+        mimic that exact structure.
+      dtype: Dtype of incoming samples and the resulting statistics.
+        By default, the dtype is `tf.float32`. Any integer dtypes will be
+        cast to corresponding floats (i.e. `tf.int32` will be cast to
+        `tf.float32`), as intermediate calculations should be performing
+        floating-point division.
+    """
+    self.shape = shape
+    self.num_chains = num_chains
+    if dtype is tf.int64:
+      dtype = tf.float64
+    elif dtype.is_integer:
+      dtype = tf.float32
+    self.dtype = dtype
+
+  def initialize(self):
+    """Initializes an empty `RunningPotentialScaleReductionState`.
+
+    Returns:
+      state: `RunningPotentialScaleReductionState` representing a stream
+        of no inputs.
+    """
+    def _initialize_for_one_state(num_chains, shape):
+      """Initializes a running variance state for one group of Markov chains."""
+      var_stream = RunningVariance(shape, self.dtype)
+      return [var_stream.initialize()
+              for _ in range(num_chains)]
+
+    chain_var = nest.map_structure_up_to(
+        self.num_chains,
+        _initialize_for_one_state,
+        self.num_chains,
+        self.shape,
+        check_types=False
+    )
+    return RunningPotentialScaleReductionState(chain_var)
+
+  def update(self, state, new_sample, chain_axis=0, chunk_axis=None):
+    """Update the `RunningPotentialScaleReductionState` with a new sample.
+
+    Args:
+      state: `RunningPotentialScaleReductionState` that represents the
+        current state of running statistics.
+      new_sample: Incoming `Tensor` sample or (possibly nested) collection of
+        `Tensor`s with shape and dtype compatible with those used to form the
+        `RunningPotentialScaleReductionState`.
+      chain_axis: The sample axis that indexes into independent Markov chains
```

Comment:
This looks different from `tfp.mcmc.potential_scale_reduction_factor`.  Is it?  Is there a reason to be different?

Reply:
It kind of is. Idea is that in the `sample_stats` version we don't force the user to conform to any sample shape semantics (i.e. [number of samples, number of chains, event shape]). Hence, the axis that specifies the independent chains can be any (I didn't think there would be a use-case where this is actually multiple axes, but again, may not have the right intuition on this).